cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,229 @@
+
+from cirq.circuits.circuit import CIRCUIT_TYPE
+from cirq.circuits import circuit, frozen_circuit
+
+
+class LogLevel(enum.Enum):
+    """Different logging resolution options for `TransformerStatsLogger`."""
+
+    INFO = 1
+    VERBOSE = 2
+    WARNING = 3
+
+
+class TransformerStatsLoggerBase:
+    """Abstract Base Class for transformer logging infrastructure.
+
+    The logger implementation should be stateful, s.t.:
+        - Each call to `register_initial` registers a new transformer stage and initial circuit.
+        - Each subsequent call to `log` should store additional logs corresponding to the stage.
+        - Each call to `register_final` should register the end of the currently active stage.
+
+    The logger assumes that
+        - Transformers are run sequentially.
+        - Nested transformers are allowed, in which case the behavior would be similar to a
+          doing a depth first search on the graph of transformers -- i.e. the top level transformer
+          would end (i.e. receive a `register_final` call) once all nested transformers (i.e. all
+          `register_initial` calls received while the top level transformer was active) have
+          finished (i.e. corresponding `register_final` calls have also been received).
+        - This behavior can be simulated by maintaining a stack of currently active stages and
+          adding data from `log` calls to the stage at the top of the stack.
+
+    The `LogLevel`s can be used to control the input processing and output resolution of the logs.
+    """
+
+    @abc.abstractmethod
+    def register_initial(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the beginning of a new transformer stage.
+
+        Args:
+            circuit: Input circuit to the new transformer stage.
+            transformer_name: Name of the new transformer stage.
+        """
+
+    @abc.abstractmethod
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        """Log additional metadata corresponding to the currently active transformer stage.
+
+        Args:
+            *args: The additional metadata to log.
+            level: Logging level to control the amount of metadata that gets put into the context.
+        """
+
+    @abc.abstractmethod
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the end of the currently active transformer stage.
+
+        Args:
+            circuit: Final transformed output circuit from the transformer stage.
+            transformer_name: Name of the (currently active) transformer stage which ends.
+        """
+
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers."""
+
+    logger: Optional[TransformerStatsLoggerBase] = None
+    ignore_tags: Tuple[Hashable, ...] = ()
+    max_parallelism: int = 1
+
+
+TRANSFORMER_TYPE = Callable[[CIRCUIT_TYPE, TransformerContext], CIRCUIT_TYPE]
```

Comment:
Since argument to a `Callable` are contravariant, defining the transformer type to be `Callable[[cirq.AbstractCircuit, TransformerContext], cirq.AbstractCircuit]` would mean that a `Callable[[cirq.Circuit, TransformerContext], cirq.Circuit]]` would not be accepted as a valid transformer. 
So, every transformer would need to accept `cirq.AbstractCircuit` and deal with all subtypes (currently, this just means calling `circuit = abstract_circuit.unfreeze()`, which can probably go in the decorator itself to avoid redundancy).  However, a much bigger issue is that every possible transformer would need to deal with any future subtypes of `AbstractCircuit` that we might end up adding, which is bad.


If we enforce that a transformer should preserve the input type, then we move the responsibility of passing in the correct type on to the user.  Though I agree that it can be annoying for the user to do the freeze/unfreeze conversions manually and we should probably consider adding it to the decorator somehow. 

What do you think?

Reply:
It's hard to decide on the best design here without seeing how people are going to use transformers, in particular how they will be composed. The main advantage I can see to carefully preserving the specific circuit types in these type signatures is that you can have the type checker tell you when calling one or another transformer that you're passing the right circuit types around and haven't forgotten a `freeze` or `unfreeze` call somewhere. But that only works if you are defining the composed transformer statically and running mypy on the code. For people composing transformers at runtime these precise types and associated type gymnastics won't help, and we'd probably want to insert runtime type checks and do the `freeze` and `unfreeze` calls to make it ergonomic for people to compose transformers (or even just to use them, for that matter). But in that case the types don't really help because it's hard for the code doing the composition to see what types were declared by the transformers it is composing (runtime reflection of mypy types is hard and not particularly well supported).

I think I'd lean toward doing the simplest thing possible by accepting and returning `cirq.AbstractCircuit`, which simplifies the API here and makes it the responsibility of the transformer implementer to call `freeze`/`unfreeze` as needed.

If we do stick with the more specific types, I'd still suggest calling this `TRANSFORMER` instead of `TRANSFORMER_TYPE`.

Reply:
How about this?

```python
TRANSFORMER = Callable[[cirq.Circuit, cirq.TransformerContext], cirq.Circuit]
TRANSFORMER_OUT = Callable[[cirq.AbstractCircuit, cirq.TransformerContext], cirq.Circuit]

def transformer(cls_or_func: TRANSFORMER) -> TRANSFORMER_OUT:
    ...
```

- Since the user is calling a "transformer" on an input circuit, it's reasonable to assume that the returned circuit type would always be a mutable circuit (similar to how calling `sorted(input_tuple)` returns a `list`). 
- The `@cirq.transformer` decorator will automatically pass `input_circuit.unfreeze(copy=True)` to the decorated transformer as input. This serves two purposes -- we avoid the redundant `unfreeze()` in each transformer and ensure that the transformer always receives a copy of the input circuit, thus ensuring that the input is never mutated (which is also supposed to be part of the new API contract). 
- A `Callable[[cirq.Circuit, TransformerContext], cirq.Circuit]]` would now be a valid transformer when type checked using the `@transformer` decorator. 
- Since the returned type is `Callable[[cirq.AbstractCircuit, cirq.TransformerContext], cirq.Circuit]`, circuit composition would now work for both frozen and mutable circuits. 


What do you think?